contributors/design-proposals/local-storage-pv.md

```diff
@@ -0,0 +1,631 @@
+
+* The paths to the mount points are always consistent, even across reboots or when storage
+is added or removed.
+* The paths are backed by a filesystem (block devices or raw partitions are not supported for
+the first phase)
+* The storage is mounted according to the provisioner discovery scheme defined below
+* The directories have appropriate permissions for the provisioner to be able to set owners and
+cleanup the volume.
+
+#### Local volume management
+
+Local PVs are statically created and not dynamically provisioned for the first phase.
+To mitigate the amount of time an administrator has to spend managing Local volumes,
+a Local static provisioner application will be provided to handle common scenarios.  For
+uncommon scenarios, a specialized provisioner can be written.
+
+The Local static provisioner will be developed in the
+[kubernetes-incubator/external-storage](https://github.com/kubernetes-incubator)
+repository, and will loosely follow the external provisioner design, with a few differences:
+
+* A provisioner instance needs to run on each node and only manage the local storage on its node.
+* For this phase, it does not handle dynamic provisioning.  Instead, it performs static provisioning
+by discovering available partitions mounted under a configured discovery directories.
+
+The basic design of the provisioner will have two separate handlers: one for PV deletion and
+cleanup, and the other for static PV creation.  A PersistentVolume informer will be created
+and its cache will be used by both handlers.
+
+PV deletion will operate on the Update event.  If the PV it provisioned changes to the “Released”
+state, and if the reclaim policy is Delete, then it will cleanup the volume and then delete the PV,
+removing it from the cache.
+
+PV creation does not operate on any informer events.  Instead, it periodically monitors the discovery
+directories, and will create a new PV for each path in the directory that is not in the PV cache.  It
+sets the "pv.kubernetes.io/provisioned-by" annotation so that it can distinguish which PVs it created.
+
+For this phase, the allowed discovery file types are directories and mount points.  The PV capacity
+will be the capacity of the underlying filesystem.  Therefore, PVs that are backed by shared
+directories will report its capacity as the entire filesystem, potentially causing overcommittment.
+Separate partitions are recommended for capacity isolation.
+
+The name of the PV needs to be unique across the cluster.  The provisioner will hash the node name,
+StorageClass name, and base file name in the volume path to generate a unique name.
+
+##### Packaging
+
+The provisioner is packaged as a container image and will run on each node in the cluster as part of
+a DaemonSet.  It needs to be run with a user or service account with the following permissions:
+
+* Create/delete/list/get PersistentVolumes - Can use the `system:persistentvolumeprovisioner` ClusterRoleBinding
```

Comment:
This gives every node write access to all PVs, which is not intended

Reply:
This is true of any external-provisioner in general.  The external provisioner design relies on the provisioner implementation to filter the PVs and only operate on PVs that it created by marking and checking the "provisioned-by" annotation.  Perhaps that design needs to be revisited so that you could create a rolebinding on a reduced set.

Reply:
Correct, but typically an external provisioner would not be required to run on all nodes, so you could avoid running it on a node containing untrusted workloads

Reply:
What about something like this: Have a single master provisioner that manages and operates on the PV objects.  Have it talk to worker pods on each node.  Each worker pod is only allowed to:

1. Report what local volumes it discovered.
2. Handle cleanup of its local volumes when requested by the master.
3. In the future, provision a new volume from a shared pool when requested by the master.

This way all the PV objects are managed by a single pod that can run on a trusted node, and each worker pod can only touch the local volumes on it's node.  The worker pods don't even need to look at or operate on PV objects.

Reply:
central management of PV resources coordinated with local management of local volumes, where the central manager only pays attention to what the local worker reports for volumes that are actually bound to the node seems better

Reply:
I updated this section to mention that it will be addressed in a future release, and added a high level overview at the bottom.